File: backend/src/main/java/com/api/bompreparo/domain/services/RecipeService.java
package com.api.bompreparo.domain.services;

import com.api.bompreparo.domain.models.Recipe;
import com.api.bompreparo.domain.models.dtos.recipe.CreateRecipeDTO;
import com.api.bompreparo.domain.models.dtos.recipe.ListRecipeDTO;
import com.api.bompreparo.domain.models.dtos.recipe.RecipeDTO;

import java.util.List;

public interface RecipeService {

    void create(Recipe obj);
    Recipe read(Long id);
    void update(Recipe obj);
    void delete(Long id);
    List<Recipe> list();

    void createRecipe(CreateRecipeDTO createRecipeDTO);
    RecipeDTO getRecipe(Long recipeId);
    void updateRecipe(Recipe recipe);
    void deleteRecipe(Long recipeId);
    List<Recipe> listRecipes();
    List<Recipe> listRecipesByCategories(List<Long> categoriesId);
    List<Recipe> listRecipesByCategory(Long categoryId);
    List<RecipeDTO> listRecipesByIngredients(List<Long> ingredientsId);
    List<ListRecipeDTO> listRecipesByUser(Long userId);

}

File: frontend/bom-preparo/src/app/domain/services/recipe.service.ts
import { Observable, map } from "rxjs";
import { RecipeRepository } from "src/app/data/repositories/recipe.repository";
import { CreateRecipeDTO } from "../models/dtos/create-recipe.dto";
import { Recipe } from "../models";
import { Injectable } from "@angular/core";
import { RecipeDTO } from "../models/dtos/recipe.dto";
import { ListRecipeDTO } from "../models/dtos/list-recipe.dto";

@Injectable({ providedIn: 'root' })
export class RecipeService {

  constructor(private recipeRepository: RecipeRepository) { }

  createRecipe(createRecipeDTO: CreateRecipeDTO): Observable<void> {
    return this.recipeRepository.createRecipe(createRecipeDTO).pipe(
      map((data) => {
        return data;
      })
    );
  }

  getRecipe(recipeId: number): Observable<RecipeDTO> {
    return this.recipeRepository.getRecipe(recipeId).pipe(
      map((data) => {
        return data;
      })
    );
  }

  listRecipes(): Observable<Recipe[]> {
    return this.recipeRepository.listRecipes().pipe(
      map((data) => {
        return data;
      })
    );
  }

  listRecipesByCategories(categoriesId: number[]): Observable<Recipe[]> {
    return this.recipeRepository.listRecipesByCategories(categoriesId).pipe(
      map((data) => {
        return data;
      })
    );
  }

  listRecipesByCategory(categoryId: number): Observable<Recipe[]> {
    return this.recipeRepository.listRecipesByCategory(categoryId).pipe(
      map((data) => {
        return data;
      })
    );
  }

  listRecipesByIngredients(ingredientsId: number[]): Observable<RecipeDTO[]> {
    return this.listRecipesByIngredients(ingredientsId).pipe(
      map((data) => {
        return data;
      })
    );
  }

  listRecipesByUser(userId: number): Observable<ListRecipeDTO[]> {
    return this.recipeRepository.listRecipesByUser(userId).pipe(
      map((data) => {
        return data;
      })
    );
  }

  updateRecipe(recipe: Recipe): Observable<void> {
    return this.recipeRepository.updateRecipe(recipe).pipe(
      map((data) => {
        return data;
      })
    );
  }

  deleteRecipe(recipeId: number): Observable<void> {
    return this.recipeRepository.deleteRecipe(recipeId).pipe(
      map((data) => {
        return data;
      })
    );
  }

}
